File: lib/export_excel.js
import fs from 'node:fs';
import * as xlsx from 'xlsx';

export function exportExcel(clients, templatePath, outputPath) {
  if (!fs.existsSync(templatePath)) {
    throw new Error(`Template file not found at ${templatePath}`);
  }

  const workbook = xlsx.readFile(templatePath);
  const sheetName = workbook.SheetNames[0];
  const sheet = workbook.Sheets[sheetName];

  // Map our keys to specific column indices in the template (0-indexed)
  const cols = [
    { dim: "Life Overall", bIdx: 3, m3Idx: 4, m6Idx: 5 },
    { dim: "Standard of Living", bIdx: 6, m3Idx: 7, m6Idx: 8 },
    { dim: "Health", bIdx: 9, m3Idx: 10, m6Idx: 11 },
    { dim: "Achieving in Life", bIdx: 12, m3Idx: 13, m6Idx: 14 },
    { dim: "Personal Relationships", bIdx: 15, m3Idx: 16, m6Idx: 17 },
    { dim: "Safety", bIdx: 18, m3Idx: 19, m6Idx: 20 },
    { dim: "Community", bIdx: 21, m3Idx: 22, m6Idx: 23 },
    { dim: "Future Security", bIdx: 24, m3Idx: 25, m6Idx: 26 },
    { dim: "Financial Worry", bIdx: 27, m3Idx: 28, m6Idx: 29 },
    { dim: "Self-Confidence", bIdx: 30, m3Idx: 31, m6Idx: 32 },
    { dim: "Voice & Agency", bIdx: 33, m3Idx: 34, m6Idx: 35 },
    { dim: "Work Readiness", bIdx: 36, m3Idx: 37, m6Idx: 38 },
    { dim: "Career Confidence", bIdx: 39, m3Idx: 40, m6Idx: 41 },
    { dim: "Skills Awareness", bIdx: 42, m3Idx: 43, m6Idx: 44 }
  ];

  let startRow = 5; // row index 5 represents Excel row 6

  for (let i = 0; i < clients.length; i++) {
    const client = clients[i];
    const rIdx = startRow + i;

    // Col A (0): Intake #
    xlsx.utils.sheet_add_aoa(sheet, [[client.intakeId]], { origin: { r: rIdx, c: 0 } });
    
    // Col B (1): Client Name
    xlsx.utils.sheet_add_aoa(sheet, [[client.id]], { origin: { r: rIdx, c: 1 } });
    
    // Col C (2): Cohort
    xlsx.utils.sheet_add_aoa(sheet, [[client.cohort]], { origin: { r: rIdx, c: 2 } });

    // Scores
    for (const mapping of cols) {
      const dim = mapping.dim;
      const bVal = client.scores.Baseline[dim] ?? null;
      const m3Val = client.scores["3mo"][dim] ?? null;
      const m6Val = client.scores["6mo"] ? (client.scores["6mo"][dim] ?? null) : null;

      if (bVal !== null) xlsx.utils.sheet_add_aoa(sheet, [[bVal]], { origin: { r: rIdx, c: mapping.bIdx } });
      if (m3Val !== null) xlsx.utils.sheet_add_aoa(sheet, [[m3Val]], { origin: { r: rIdx, c: mapping.m3Idx } });
      if (m6Val !== null) xlsx.utils.sheet_add_aoa(sheet, [[m6Val]], { origin: { r: rIdx, c: mapping.m6Idx } });
    }
  }

  xlsx.writeFile(workbook, outputPath);
  return { success: true };
}

File: lib/export_excel.py
# lib/export_excel.py
import sys
import json
import openpyxl
import os

def export_excel(input_json_path, template_path, output_path):
    try:
        # Load clients data
        with open(input_json_path, 'r') as f:
            clients = json.load(f)
            
        # Open the template workbook
        wb = openpyxl.load_workbook(template_path)
        
        # Ensure "Data Entry" sheet exists
        if "Data Entry" not in wb.sheetnames:
            raise Exception("Template missing 'Data Entry' sheet")
            
        sheet = wb["Data Entry"]
        
        # Clear existing data rows (from row 6 to max_row)
        # We don't delete rows entirely to preserve formatting, we just clear values
        for row_idx in range(6, max(6, sheet.max_row + 1)):
            for col_idx in range(1, 46):
                sheet.cell(row=row_idx, column=col_idx).value = None

        # Dimension to column offset mapping
        dimensions_map = {
            "Life Overall": 4,
            "Standard of Living": 7,
            "Health": 10,
            "Achieving in Life": 13,
            "Personal Relationships": 16,
            "Safety": 19,
            "Community": 22,
            "Future Security": 25,
            "Financial Worry": 28,
            "Self-Confidence": 31,
            "Voice & Agency": 34,
            "Work Readiness": 37,
            "Career Confidence": 40,
            "Skills Awareness": 43
        }
        
        # Populate the template
        start_row = 6
        for i, client in enumerate(clients):
            row_idx = start_row + i
            
            # Basic info
            sheet.cell(row=row_idx, column=1).value = client.get("intakeId")
            sheet.cell(row=row_idx, column=2).value = client.get("id")
            sheet.cell(row=row_idx, column=3).value = client.get("cohort")
            
            # Scores
            scores = client.get("scores", {})
            baseline = scores.get("Baseline", {}) or {}
            mo3 = scores.get("3mo", {}) or {}
            mo6 = scores.get("6mo", {}) or {}
            
            for dim, offset in dimensions_map.items():
                if dim in baseline and baseline[dim] is not None:
                    sheet.cell(row=row_idx, column=offset).value = baseline[dim]
                if dim in mo3 and mo3[dim] is not None:
                    sheet.cell(row=row_idx, column=offset + 1).value = mo3[dim]
                if mo6 and dim in mo6 and mo6[dim] is not None:
                    sheet.cell(row=row_idx, column=offset + 2).value = mo6[dim]
                    
        # Save to output path
        wb.save(output_path)
        
        return {"ok": True, "file": output_path}
        
    except Exception as e:
        return {"error": str(e)}

if __name__ == "__main__":
    if len(sys.argv) < 4:
        print(json.dumps({"error": "Missing arguments. Usage: python3 export_excel.py <input_json> <template_xlsx> <output_xlsx>"}))
        sys.exit(1)
        
    input_json = sys.argv[1]
    template = sys.argv[2]
    output = sys.argv[3]
    
    result = export_excel(input_json, template, output)
    print(json.dumps(result))
